packages/test/test-utils/src/opProcessingController.ts

```diff
@@ -75,13 +80,44 @@ class DeltaManagerMonitor extends DeltaManagerToggle {
     private pendingCount: number = 0;
     private clientId: string | undefined;
     private firstClientSequenceNumber: number = -1;
-    private _latestSequenceNumber: number;
+    private lastOutbound: IDocumentMessage | undefined;
+    private readonly lastInboundPerClient = new Map<string, ISequencedDocumentMessage>();
+    private pendingWriteConnection = false;
+
+    public expectingInboundFrom(outbound: DeltaManagerMonitor): boolean {
+        // there should be no outstanding work for disposed delta managers
+        if (this.deltaManager.disposed || outbound.deltaManager.disposed) {
+            return false;
+        }
+        // if there is no last outbound, we are not waiting for anything
+        if (outbound.lastOutbound === undefined
+            || outbound.clientId === undefined) {
+            return false;
+        }
+        // if out inbound is paused we are not expecting to receive anything more
+        if (this.inboundPaused) {
+            return false;
+        }
+
+        // if outbound is ourself, return if we having pending work
+        if (this === outbound) {
+            return this.hasPendingWork();
+        }
+
+        // check if we are waiting to see a message from outbound
+        const lastInboundForOutbound = this.lastInboundPerClient.get(outbound.clientId);
```

Comment:
This assumes there are no disconnects, i.e. clientId does not change through tests, right? Is this important condition? I.e. should this class assert if disconnects happened (as part of UT mimicking disconnection logic) as this code would start lying in some cases?

Reply:
we clear lastOutbound on disconnect. i think this can lead to missing outbound work if the sending client disconnects with ops in flight, but that was already the case for this class. i don't think we can assert on disconnects, as this should work against real servers, so disconnects can happen